### PR TITLE
Add protocol sdk and minting features

### PR DESCRIPTION
**Pull request by iainnash**
Add new functions to the `premint-sdk` and rename it to `protocol-sdk`.

### Mint token (not premint):

(Ideally this can also handle premints but the data paths are too different for now).

```ts
const { send } = await minter.mintToken({
  address: targetContract,
  tokenId: targetTokenId,
  publicClient,
  sender: creatorAccount,
  mintArguments: {
    mintToAddress: creatorAccount,
    quantityToMint: 1,
  },
});
```

### Create token/contract:

```ts
const new1155TokenRequest = await createNew1155Token({
  publicClient,
  contract: {
    name: "testContract",
    uri: demoContractMetadataURI,
  },
  tokenMetadataURI: demoTokenMetadataURI,
  account: creatorAccount,
  mintToCreatorCount: 1,
});
```

### CODE REVIEW

**Comment by iainnash** [comment]
Current dependencies on/for this PR:
* main
  * **PR #322** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/322" target="_blank">Graphite</a>  👈
    * **PR #324** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/324" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ourzora/zora-protocol/322?utm_source=stack-comment).

File: packages/protocol-sdk/src/1155-create-helper.ts [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+  salesConfig: SalesConfigParamsType;
+  royaltySettings?: RoyaltySettingsType;
+}) {
+  if (!maxSupply) {
+    maxSupply = OPEN_EDITION_MINT_SIZE;
+  }
+  maxSupply = BigInt(maxSupply);
+  mintToCreatorCount = BigInt(mintToCreatorCount);
+
+  const salesConfigWithDefaults = Object.assign(
```
Comment (oveddan) [inline packages/protocol-sdk/src/1155-create-helper.ts:80]: you can do:
```ts
{
    ...DEFAULT_SALE_SETTINGS,
    ...salesConfig
}
```

File: packages/protocol-sdk/src/1155-create-helper.ts [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+      defaultAdmin?: Address;
+    }
+  | Address;
+
+type RoyaltySettingsType = {
+  royaltyBPS: number;
+  royaltyRecipient: Address;
+};
+
+export function create1155TokenSetupArgs({
```
Comment (oveddan) [inline packages/protocol-sdk/src/1155-create-helper.ts:53]: this is more around creating an 1155 using the fixed priced minter.  do we want to have it's name to reflect that?  cause this naming implies that this is the only way to create an 1155 token

File: packages/protocol-sdk/src/1155-create-helper.ts [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+const royaltyBPSDefault = 1000;
+
+type SalesConfigParamsType = {
+  // defaults to 0
+  pricePerToken?: bigint;
+  // defaults to 0, in seconds
+  saleStart?: bigint;
+  // defaults to forever, in seconds
+  saleEnd?: bigint;
+  maxTokensPerAddress?: number;
```
Comment (oveddan) [inline packages/protocol-sdk/src/1155-create-helper.ts:27]: could you document here that 0n means no limit?

File: packages/protocol-sdk/src/1155-create-helper.ts [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+  fundsRecipient?: Address;
```
Comment (oveddan) [inline packages/protocol-sdk/src/1155-create-helper.ts:28]: please add explanation that this is just for paid mints, and what id defaults to (if any)

File: packages/protocol-sdk/src/1155-create-helper.ts [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+  mintToCreatorCount,
+  tokenMetadataURI,
+  fixedPriceMinterAddress,
+  createReferral,
+  maxSupply,
+  account,
+  salesConfig,
+  royaltySettings,
+}: {
+  maxSupply?: bigint | number;
```
Comment (oveddan) [inline packages/protocol-sdk/src/1155-create-helper.ts:64]: how about just requiring a bigint here and making things simpler? (prevent it from being a number)

File: packages/protocol-sdk/src/1155-create-helper.ts [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+  nextTokenId: bigint;
+  mintToCreatorCount: bigint | number;
+  account: Address;
+  tokenMetadataURI: string;
+  fixedPriceMinterAddress: Address;
+  salesConfig: SalesConfigParamsType;
+  royaltySettings?: RoyaltySettingsType;
+}) {
+  if (!maxSupply) {
+    maxSupply = OPEN_EDITION_MINT_SIZE;
```
Comment (oveddan) [inline packages/protocol-sdk/src/1155-create-helper.ts:75]: you can do defaults like:

```ts
export function create1155TokenSetupArgs({
   maxSupply = OPEN_EDITION_MINT_SIZE
}: {
   maxSupply?: bigint | numer;
})

```

File: packages/protocol-sdk/src/1155-create-helper.ts [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+
+  const salesConfigWithDefaults = Object.assign(
+    // Set static sales default.
+    DEFAULT_SALE_SETTINGS,
+    // Override with user settings.
+    salesConfig,
+  );
+
+  const setupActions = [
+    encodeFunctionData({
```
Comment (oveddan) [inline packages/protocol-sdk/src/1155-create-helper.ts:88]: could you extract these to functions with names explaining what they do, like:

```ts
const setupActions = [
    addFixedPricedMinterMintingPermissios({fixedPricedMinterAddress}),
    validateNextTokenId({nextTokenId}),
    setupNewToken({createReferral, tokenMetadataURI, maxSupply})
    ...etc
]
```

this would help reduce the number of lines in this parent function, and make it more readable, also bringing composability/reusability to these setup actions

File: packages/protocol-sdk/src/1155-create-helper.ts [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+          },
+        ],
+      }),
+    );
+  }
+
+  return setupActions;
+}
+
+export const getTokenIdFromCreateReceipt = (
```
Comment (oveddan) [inline packages/protocol-sdk/src/1155-create-helper.ts:154]: very useful function, thx for adding.

File: packages/protocol-sdk/src/1155-create-helper.ts [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+  account: Address,
+) {
+  let contractAddress;
+  let contractExists = false;
+  if (typeof contract !== "string") {
+    contractAddress = await publicClient.readContract({
+      abi: zoraCreator1155FactoryImplABI,
+      // Since this address is deterministic we can hardcode a chain id safely here.
+      address: zoraCreator1155FactoryImplAddress[999],
+      functionName: "deterministicContractAddress",
```
Comment (oveddan) [inline packages/protocol-sdk/src/1155-create-helper.ts:183]: this function only works for contracts created via determinstic contract addresses, i.e. premint.

File: packages/protocol-sdk/src/1155-create-helper.ts [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+        ...data,
+      });
+      if (decodedLog && decodedLog.eventName === "SetupNewToken") {
+        return decodedLog.args.tokenId;
+      }
+    } catch (err: any) {}
+  }
+};
+
+async function getContractExists(
```
Comment (oveddan) [inline packages/protocol-sdk/src/1155-create-helper.ts:171]: i think it may be cleaner to just break this into two functions:

```ts
const getDeterminsticContractExists({contractConfig})

const getContractExists({address})
```

File: packages/protocol-sdk/src/1155-create-helper.ts [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+  tokenMetadataURI,
+  mintToCreatorCount = 1,
+  salesConfig = {},
+  maxSupply,
+  account,
+  royaltySettings,
+  getAdditionalSetupActions,
+}: {
+  publicClient: PublicClient;
+  account: Address;
```
Comment (oveddan) [inline packages/protocol-sdk/src/1155-create-helper.ts:223]: what account does this represent? could you add a tsdoc comment explaining?

File: packages/protocol-sdk/src/1155-create-helper.ts [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+      ...getAdditionalSetupActions({ tokenId: nextTokenId, contractAddress }),
+      ...tokenSetupActions,
+    ];
+  }
+
+  if (!contractAddress && typeof contract === "string") {
+    throw new Error("Invariant: contract cannot be missing and an address");
+  }
+  if (!contractExists && typeof contract !== "string") {
+    const { request } = await publicClient.simulateContract({
```
Comment (oveddan) [inline packages/protocol-sdk/src/1155-create-helper.ts:279]: this function is very useful, but it is doing a bit too much, reflected in the long length; id actually break this down into a less opinionated function that returns params that you can submit to viem or wagmi or even ethers.js

Like:

`makePrepareNewToken` - this creates a resulting config object that, and can be plugged into viem or wagmi hooks:

```ts
// makePrepareNewToken is sdk function
const prepareCreateNewTokenParams = makePrepareNewToken({
  maxSupply,
  account,
  royaltySettings,
  getAdditionalSetupActions,
  ...etc
});

// with viem
const { request } = await publicClient.simulateContract(prepareCreateNewTokenParams):

// with wagmi
const { config, error } = usePrepareContractWrite(prepareCreateNewTokenParams);
```

File: packages/protocol-sdk/src/constants.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+// Subgraph base settings
+const CONFIG_BASE =
+  "https://api.goldsky.com/api/public/project_clhk16b61ay9t49vm6ntn4mkz/subgraphs";
+
+function getSubgraph(name: string, version: string): string {
+  return `${CONFIG_BASE}/${name}/${version}/gn`;
+}
+
+export const ZORA_SUBGRAPH_URLS: Record<number, string> = {
```
Comment (oveddan) [inline packages/protocol-sdk/src/constants.ts:12]: do you thin the subgraph stuff can be broken into another PR? the pr is getting kind of big to review

**Review by oveddan — CHANGES_REQUESTED** [review]

File: packages/protocol-sdk/src/client-base.ts [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    this.network = networkConfig;
+  }
+
+  /**
+   * Getter for public client that instantiates a publicClient as needed
+   *
+   * @param publicClient Optional viem public client
+   * @returns Existing public client or makes a new one for the given chain as needed.
+   */
+  protected getPublicClient(publicClient?: PublicClient): PublicClient {
```
Comment (oveddan) [inline packages/protocol-sdk/src/client-base.ts:65]: this can be extracted to a static utility function:

`getPublicClient(publicClient?: PublicClient, chain: Chain)` and then we don't need the inheritance anymore

File: packages/protocol-sdk/src/client-base.ts [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+  },
+  [foundry.id]: {
+    chainId: foundry.id,
+    isTestnet: true,
+    zoraPathChainName: "zgor",
+    zoraBackendChainName: BackendChainNamesLookup.ZORA_GOERLI,
+  },
+};
+
+export abstract class ClientBase {
```
Comment (oveddan) [inline packages/protocol-sdk/src/client-base.ts:46]: id avoid having this base class just for holding the chain and network config.  how about just having the subclasses contain these objects and have utility functions to extract whats needed from both objects?

inheritance can be kind of hard to follow, and id be concerned that more stuff gets thrown in here than is needed.

File: packages/protocol-sdk/src/mint-client.ts [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+            mintArguments.mintComment || "",
+          ]),
+          mintArguments.mintReferral || zeroAddress,
+        ],
+      });
+      return await walletClient.writeContract(request);
+    };
+  }
+
+  private async mintZora721({
```
Comment (oveddan) [inline packages/protocol-sdk/src/mint-client.ts:178]: this PR is a bit big to review, could you break out the mint functionality into its own PR (separate from creation of tokens)

**Review by oveddan — CHANGES_REQUESTED** [review]

File: packages/protocol-sdk/README.md
[the remark this answers is not in the document]
Comment (iainnash) [reply packages/protocol-sdk/README.md:1]: I did but I think i edited it too much.

File: packages/protocol-sdk/src/1155-create-helper.ts (hunk 9 above)
Comment (iainnash) [reply packages/protocol-sdk/src/1155-create-helper.ts]: We're only using this for when people don't pass in a contract address. I'll document it.

File: packages/protocol-sdk/src/1155-create-helper.ts (hunk 12 above)
Comment (iainnash) [reply packages/protocol-sdk/src/1155-create-helper.ts]: Ran into type issues – waiting to chat about a fix.

File: packages/protocol-sdk/src/constants.ts (hunk 13 above)
Comment (iainnash) [reply packages/protocol-sdk/src/constants.ts]: It's kinda needed here – I broke out some other things. Since it's new code I think it's okay for now?

Comment (oveddan) [reply packages/protocol-sdk/src/constants.ts]: yeah i realize its fine, sounds good

File: packages/protocol-sdk/src/client-base.ts (hunk 15 above)
Comment (oveddan) [reply packages/protocol-sdk/src/client-base.ts]: fine resolving this in a future PR (to get this out)

Comment (iainnash) [reply packages/protocol-sdk/src/client-base.ts]: The main reason I have things setup this way is the ease to override a network client without adding a ton of parameters.

**Review by oveddan — APPROVED** [review]
